### PR TITLE
Implement no-copy subarray iteration (Implements #3386)

### PR DESCRIPTION
**Pull request by cjgajard**
Since I deleted my previous fork I need to PR again.
This PR closes #3390.

### CODE REVIEW

File: src/indexable.cr [hunk 1]
```diff
@@ -165,6 +165,51 @@ module Indexable(T)
...
+  # array = ["a", "b", "c", "d", "e"]
+  # array.each(1..3) { |x| print x, " -- " }
+  # ```
+  #
+  # produces:
+  #
+  # ```text
+  # b -- c -- d --
+  # ```
+  def each(*, within range : Range(Int, Int))
```
Comment (oprypin) [inline src/indexable.cr]: I don't really like requiring `within:`. Even the example doesn't have it and so appears broken 😛

Comment (cjgajard) [inline src/indexable.cr]: Tank you for checking, the example is broken. The start/count overload also use required named-arguments, I think changing the example should be enough

**Comment by akzhan** [comment]
LGTM

File: src/indexable.cr [hunk 2]
```diff
@@ -201,6 +246,41 @@ module Indexable(T)
...
+  #
+  # Negative indices count backward from the end of the array. (-1 is the
+  # last element).
+  #
+  # Raises `IndexError` if the starting index is out of range.
+  # Raises `ArgumentError` if `count` is a negative number.
+  #
+  # ```
+  # array = ["a", "b", "c", "d", "e"]
+  # array.each(start: 1, count: 3) { |x| print x, " -- " }
```
Comment (Sija) [inline src/indexable.cr]: should be `array.each_index`.

**Comment by sdogruyol** [comment]
@cjgajard this looks good :+1: Are you around to fix the minor issues?

**Comment by RX14** [comment]
They should at least be forced for start and count, and I don't mind the range being a forced named parameter too. I like the named params

**Comment by yxhuvud** [comment]
@RX14 A better question is why count is mandatory. Default could be to iterate to end.

**Comment by RX14** [comment]
You could make it default to `-1`... I don't really care either way.

File: src/indexable.cr [hunk 3]
```diff
@@ -201,6 +246,41 @@ module Indexable(T)
...
+  # array.each_index(start: -3, count: 2) { |x| print x, " -- " }
+  # ```
+  #
+  # produces:
+  #
+  # ```text
+  # 2 -- 3 --
+  # ```
+  def each_index(*, start : Int, count : Int)
+    raise ArgumentError.new "negative count: #{count}" if count < 0
```
Comment (Sija) [inline src/indexable.cr]: Oops, _Title cased_ `Exception` messages, please.

**Pull request merged**